egs/aishell/s10/chain/egs_dataset.py

```diff
@@ -0,0 +1,206 @@
+
+
+def get_egs_dataloader(egs_dir, egs_left_context, egs_right_context):
+
+    dataset = NnetChainExampleDataset(egs_dir=egs_dir)
+    frame_subsampling_factor = 3
+
+    # we have merged egs offline, so batch size is 1
+    batch_size = 1
+
+    collate_fn = NnetChainExampleDatasetCollateFunc(
+        egs_left_context=egs_left_context,
+        egs_right_context=egs_right_context,
+        frame_subsampling_factor=frame_subsampling_factor)
+
+    dataloader = DataLoader(dataset,
+                            batch_size=batch_size,
+                            num_workers=0,
+                            collate_fn=collate_fn)
+    return dataloader
+
+
+def read_nnet_chain_example(rxfilename):
+    eg = nnet3.NnetChainExample()
+    ki = kaldi.Input(rxfilename=rxfilename)
+    eg.Read(ki.Stream(), True)
+    ki.Close()
+    return eg
+
+
+class NnetChainExampleDataset(Dataset):
+
+    def __init__(self, egs_dir):
+        '''
+        We assume that there exist many cegs.*.scp files inside egs_dir
+        '''
+        self.scps = glob.glob('{}/cegs.*.scp'.format(egs_dir))
+        assert len(self.scps) > 0
+        self.items = list()
+        for scp in self.scps:
+            with open(scp, 'r') as f:
+                for line in f:
+                    # line should be: "key rxfilename"
+                    split = line.split()
+                    assert len(split) == 2
+                    self.items.append(split)
+
+    def __len__(self):
+        return len(self.items)
+
+    def __getitem__(self, i):
+        return self.items[i]
+
+    def __str__(self):
+        s = 'num egs scps: {}\n'.format(len(self.scps))
+        s = 'num egs: {}\n'.format(len(self.items))
+        return s
+
+
+class NnetChainExampleDatasetCollateFunc:
+
+    def __init__(self, egs_left_context, egs_right_context,
+                 frame_subsampling_factor):
+        '''
+        egs_left_context is from egs/info/left_context
+        egs_right_context is from egs/info/right_context
+        '''
+
+        assert egs_left_context >= 0
+        assert egs_right_context >= 0
+
+        # we currently support either no subsampling or
+        # subsampling factor to be 3
+        assert frame_subsampling_factor in [1, 3]
+
+        self.egs_left_context = egs_left_context
+        self.egs_right_context = egs_right_context
+        self.frame_subsampling_factor = frame_subsampling_factor
+
+    def __call__(self, batch):
+        '''
+        batch is a list of [key, rxfilename]
+        returned from `__getitem__()` of `NnetChainExampleDataset`
+
+        Since we combined egs offline, the batch size is usually one.
+        '''
+
+        key_list = []
+
+        # contains a list of a 3D array
+        # of shape [batch_size, seq_len, feat_dim]
+        feature_list = []
+
+        supervision_list = []
+        for b in batch:
+            key, rxfilename = b
+            key_list.append(key)
+            eg = read_nnet_chain_example(rxfilename)
```

Comment:
Hm... I'm a bit concerned about speed, here.   I think it would  be preferable to have an iterable-style dataset.  The issue is that it will repeatedly have to do fopen(), fseek(), fclose(), while if it were just reading the archive sequentially it would be just read() each time.

Reply:
Something else I should point out is that with the newer 'chain2'-style egs dir (to be merged), we rely more on scp files, which makes it easier to randomize things and combine previously dumped egs.  But it's still important to use a SequentialTableReader because we go to a lot of trouble to make sure that most of the reads are sequential (the scp entries mostly point to successive egs in an archive).